pkg/operator/builder.go

```diff
@@ -0,0 +1,337 @@
+	b.namespaces = namespaces
+	return b
+}
+
+func (b *Builder) WithLogger(logger *zap.Logger) *Builder {
+	b.logger = logger
+	return b
+}
+
+func (b *Builder) WithSyncPeriod(period time.Duration) *Builder {
+	b.syncPeriod = period
+	return b
+}
+
+func (b *Builder) WithMetricAddress(address string) *Builder {
+	b.metricAddress = address
+	return b
+}
+
+func (b *Builder) WithProbeAddress(address string) *Builder {
+	b.probeAddress = address
+	return b
+}
+
+func (b *Builder) WithLeaderElection(enable bool) *Builder {
+	b.leaderElection = enable
+	return b
+}
+
+func (b *Builder) WithAtlasDomain(domain string) *Builder {
+	b.atlasDomain = domain
+	return b
+}
+
+func (b *Builder) WithPredicates(predicates []predicate.Predicate) *Builder {
+	b.predicates = predicates
+	return b
+}
+
+func (b *Builder) WithAPISecret(apiSecret client.ObjectKey) *Builder {
+	b.apiSecret = apiSecret
+	return b
+}
+
+func (b *Builder) WithAtlasProvider(provider atlas.Provider) *Builder {
+	b.atlasProvider = provider
+	return b
+}
+
+func (b *Builder) WithFeatureFlags(featureFlags *featureflags.FeatureFlags) *Builder {
+	b.featureFlags = featureFlags
+	return b
+}
+
+func (b *Builder) WithDeletionProtection(deletionProtection bool) *Builder {
+	b.deletionProtection = deletionProtection
+	return b
+}
+
+// Build builds the controller manager and configure operator controllers
+func (b *Builder) Build(ctx context.Context) (manager.Manager, error) {
+	mergeDefaults(b)
+
+	cacheOpts := cache.Options{
+		SyncPeriod: &b.syncPeriod,
+	}
+
+	if len(b.namespaces) == 0 {
+		cacheOpts.ByObject = map[client.Object]cache.ByObject{
+			&corev1.Secret{}: {
+				Label: labels.SelectorFromSet(labels.Set{
+					connectionsecret.TypeLabelKey: connectionsecret.CredLabelVal,
+				}),
+			},
+		}
+	} else {
+		cacheOpts.DefaultNamespaces = map[string]cache.Config{}
+		for _, namespace := range b.namespaces {
+			cacheOpts.DefaultNamespaces[namespace] = cache.Config{}
+		}
+	}
+
+	mgr, err := b.managerProvider.New(
+		b.config,
+		ctrl.Options{
+			Scheme:  b.scheme,
+			Metrics: metricsserver.Options{BindAddress: b.metricAddress},
+			WebhookServer: webhook.NewServer(webhook.Options{
+				Port: 9443,
+			}),
+			Cache:                  cacheOpts,
+			HealthProbeBindAddress: b.probeAddress,
+			LeaderElection:         b.leaderElection,
+			LeaderElectionID:       b.leaderElectionID,
+		},
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err = mgr.AddHealthzCheck("health", healthz.Ping); err != nil {
+		return nil, err
+	}
+
+	if err = mgr.AddReadyzCheck("check", healthz.Ping); err != nil {
+		return nil, err
+	}
+
+	if err = indexer.RegisterAll(ctx, mgr, b.logger); err != nil {
+		return nil, fmt.Errorf("unable to create indexers: %w", err)
+	}
+
+	if b.atlasProvider == nil {
+		b.atlasProvider = atlas.NewProductionProvider(b.atlasDomain, b.apiSecret, mgr.GetClient())
+	}
+
+	_, err = atlasproject.NewAtlasProjectReconciler(
```

Comment:
We are ignoring the reconciler everywhere, why not instead this?

```suggestion
	err = atlasproject.RegisterAtlasProjectReconciler(
```

It is basically registering the reconciler within the manager right?

Reply:
it does construct the reconciler, register it tho the manager is a bonus(if we can call it like this)

Reply:
Yes but my point is that the main usage of these functions do NOT use the returned reconcilers so, why do we return it at all in the signature?

I mean there seems to be a API impedance mismatch here.